File: solutions/234.palindrome-linked-list/palindrome-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    let preV = null;
    let cur = head;
    let curV =( cur === null ? null : new ListNode(cur.val));
    while(cur !== null){
        cur = cur.next;
        const node = (cur === null ? null : new ListNode(cur.val));
        curV.next = preV;
        preV = curV;
        curV = node;
    }
    cur = head;
    while(preV !== null){
        if(preV.val !== cur.val){
            return false;
        }
        preV = preV.next;
        cur = cur.next;
    }
    return true;
};

File: solutions/234.palindrome-linked-list/palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        
        // get length 
        ListNode cur = head;
        int len = 0;
        while(cur != null){
            len++;
            cur = cur.next;
        }
        // get mid index
        int mid = (len + 1) / 2;
        cur = head;
        while(mid > 0){
            cur = cur.next;
            mid--;
        }
        // reverse palindrome part
        ListNode pre = null;
        while(cur != null){
            ListNode temp = cur.next;
            cur.next = pre;
            pre = cur;
            cur = temp;
        }
        // compare 
        cur = head;
        while(pre != null){
            if(pre.val != cur.val) return false;
            pre = pre.next;
            cur = cur.next;
        }
        return true;       
    }
}